File: julia/Day19.jl
module Day19

export Submarine, on_input!, output

using LinearAlgebra

const forward_directions = [
    [1, 0, 0],
    [-1, 0, 0],
    [0, 1, 0],
    [0, -1, 0],
    [0, 0, 1],
    [0, 0, -1]
]

function build_possible_rotation_matrices()::Array{Int32}
    result = Array{Int32}(undef, (3, 3, 0))

    for forward_i = 1:6
        x = forward_directions[forward_i]

        y_possibilities = [
            [x[3], x[1], x[2]],
            [-x[3], -x[1], -x[2]],
            [x[2], x[3], x[1]],
            [-x[2], -x[3], -x[1]],
        ]

        for y in y_possibilities
            z = x × y

            rot_mat = [x y z]
            result = [result;;; rot_mat]
        end
    end

    result
end

mutable struct Submarine
    scanner_reports::Vector{Matrix{Int32}}
    beacons_relative_to_0::Set{Vector{Int32}}
    scanner_position_relative_to_0::Matrix{Int32}

    scanner_possible_rotations::Array{Int32}

    Submarine() = new(Vector{Matrix{Int32}}(), Set{Vector{Int32}}(), Matrix(undef, 3, 0), build_possible_rotation_matrices())
end

function on_input!(s::Submarine, a::AbstractString)
    if startswith(a, "---")
        push!(s.scanner_reports, Matrix(undef, 3, 0))
        s.scanner_position_relative_to_0 = [s.scanner_position_relative_to_0 [0, 0, 0]]
        return
    end

    if isempty(a)
        return
    end

    coords = parse.(Int32, split(a, ","))
    s.scanner_reports[end] = [s.scanner_reports[end] coords]
end

function output(s::Submarine)
    triangulate(s)

    total_beacons = length(s.beacons_relative_to_0)

    println("Part 1: $(total_beacons)")
    println("Part 2: $(max_manhattan_between_scanners(s))")
end

function manhattan(p1::Vector{Int32}, p2::Vector{Int32})::Int32
    sum(abs.(p1 - p2))
end

function max_manhattan_between_scanners(s::Submarine)::Int32
    result::Int32 = 0
    total_scanners = length(s.scanner_reports)

    for i in 1:total_scanners
        for j in i + 1:total_scanners
            result = max(result, manhattan(s.scanner_position_relative_to_0[:, i], s.scanner_position_relative_to_0[:, j]))
        end
    end

    result
end

function invert_tuple(t::Tuple{Int32,Int32})::Tuple{Int32,Int32}
    (t[2], t[1])
end

function triangulate_scanners(s::Submarine, i, j)::Bool
    if !iszero(s.scanner_position_relative_to_0[:, i]) && !iszero(s.scanner_position_relative_to_0[:, j])
        return false
    end

    common = common_beacons(s.scanner_reports[i], s.scanner_reports[j])

    if length(common) < 12
        return false
    end

    from = i
    to = j

    if !iszero(s.scanner_position_relative_to_0[:, to])
        from, to = to, from

        common = map(invert_tuple, common)
    end

    total_possible_rotations = size(s.scanner_possible_rotations)[3]

    for rotation = 1:total_possible_rotations
        adjusted_points = s.scanner_possible_rotations[:, :, rotation] * s.scanner_reports[to]

        dists = collect(map(x -> s.scanner_reports[from][:, x[1]] - adjusted_points[:, x[2]], common))
        if all(x -> x == dists[1], dists)
            s.scanner_reports[to] = adjusted_points
            s.scanner_position_relative_to_0[:, to] = s.scanner_position_relative_to_0[:, from] + dists[1]

            for p in eachcol(adjusted_points)
                push!(s.beacons_relative_to_0, p + s.scanner_position_relative_to_0[:, to])
            end

            return true
        end
    end

    return false
end

function triangulate(s::Submarine)
    total_scanners = length(s.scanner_reports)

    for p in eachcol(s.scanner_reports[1])
        push!(s.beacons_relative_to_0, p)
    end

    visited = falses(total_scanners)
    visited[1] = true

    for i = 2:total_scanners
        if triangulate_scanners(s, 1, i)
            visited[i] = true
        end
    end

    while !all(visited)
        for i = 2:total_scanners
            if !visited[i]
                continue
            end

            for j = 2:total_scanners
                if visited[j]
                    continue
                end

                if triangulate_scanners(s, i, j)
                    visited[j] = true
                end
            end
        end
    end
end

function squared_dist(p1::Vector{Int32}, p2::Vector{Int32})::Int32
    sum((p1 - p2) .^ 2)
end

# The following will allow us to compare tuples only looking at the first element, which is what we want for the code below.
import Base
# Sets in julia are backed by dictionaries, so we need to override the `==` and `isequal` behaviors for some set operations, as well as override the `hash` behavior for the dictionary.
Base.:(==)(lhs::Tuple{Int32,Int32}, rhs::Tuple{Int32,Int32}) = (lhs[1] == rhs[1])
Base.isequal(x::Tuple{Int32,Int32}, y::Tuple{Int32,Int32}) = x == y
Base.hash(x::Tuple{Int32,Int32}, h::UInt) = hash(x[1], h)

function squared_distance_sets(beacons::Matrix{Int32})::Vector{Set{Tuple{Int32,Int32}}}
    n = size(beacons)[2]

    result = [Set{Tuple{Int32,Int32}}() for _ = 1:n]

    for i::Int32 in 1:n
        for j::Int32 in i+1:n
            dist = squared_dist(beacons[:, i], beacons[:, j])
            push!(result[i], (dist, j))
            push!(result[j], (dist, i))
        end
    end

    result
end

function common_beacons(beacons1::Matrix{Int32}, beacons2::Matrix{Int32})::Vector{Tuple{Int32,Int32}}
    result = Vector{Tuple{Int32,Int32}}()

    dist1 = squared_distance_sets(beacons1)
    dist2 = squared_distance_sets(beacons2)

    for i = 1:length(dist1)
        for j = 1:length(dist2)
            common_dists = dist1[i] ∩ dist2[j]

            if length(common_dists) < 11
                continue
            end

            push!(result, (i, j))

            # We'll use this to get the actual tuple in each set because we can't do that directly in sets.
            collected_dist1 = collect(dist1[i])
            collected_dist2 = collect(dist2[j])

            for d in common_dists
                pos1 = collected_dist1[findfirst(isequal(d), collected_dist1)]
                pos2 = collected_dist2[findfirst(isequal(d), collected_dist2)]

                push!(result, (pos1[2], pos2[2]))
            end
        end
    end

    result
end

end

File: 2021/rust/src/day19.rs
use std::{collections::BTreeSet};

// All 24 possible rotation states that a scanner can be in are listed here. Each function will translate between the scanner's local coordinate system to a "global" coordinate system. I calculated all of these just using my hand (right-hand rule, although I used my left hand, but it doesn't really matter which hand you use) and going through each case one by one and listing the directions.
const ROTATIONS: [fn(&Point) -> Point; 24] = [
    |p: &Point| Point { x: p.x, y: p.y, z: p.z },
    |p: &Point| Point { x: p.x, y: -p.z, z: p.y },
    |p: &Point| Point { x: p.x, y: -p.y, z: -p.z },
    |p: &Point| Point { x: p.x, y: p.z, z: -p.y },

    |p: &Point| Point { x: -p.x, y: -p.y, z: p.z },
    |p: &Point| Point { x: -p.x, y: -p.z, z: -p.y },
    |p: &Point| Point { x: -p.x, y: p.y, z: -p.z },
    |p: &Point| Point { x: -p.x, y: p.z, z: p.y },

    |p: &Point| Point { x: -p.y, y: p.x, z: p.z },
    |p: &Point| Point { x: p.z, y: p.x, z: p.y },
    |p: &Point| Point { x: -p.z, y: p.x, z: -p.y },
    |p: &Point| Point { x: p.y, y: p.x, z: -p.z },

    |p: &Point| Point { x: -p.y, y: -p.x, z: -p.z },
    |p: &Point| Point { x: -p.z, y: -p.x, z: p.y },
    |p: &Point| Point { x: p.y, y: -p.x, z: p.z },
    |p: &Point| Point { x: p.z, y: -p.x, z: -p.y },

    |p: &Point| Point { x: -p.z, y: p.y, z: p.x },
    |p: &Point| Point { x: p.y, y: p.z, z: p.x },
    |p: &Point| Point { x: -p.y, y: -p.z, z: p.x },
    |p: &Point| Point { x: p.z, y: -p.y, z: p.x },

    |p: &Point| Point { x: p.z, y: p.y, z: -p.x },
    |p: &Point| Point { x: -p.y, y: p.z, z: -p.x },
    |p: &Point| Point { x: -p.z, y: -p.y, z: -p.x },
    |p: &Point| Point { x: p.y, y: -p.z, z: -p.x },
];

#[derive(Clone, Copy, Eq, Ord, PartialEq, PartialOrd)]
struct Point {
    x: i32,
    y: i32,
    z: i32,
}

impl std::ops::Sub for &Point {
    type Output = Point;

    fn sub(self, rhs: Self) -> Self::Output {
        Point {
            x: self.x - rhs.x,
            y: self.y - rhs.y,
            z: self.z - rhs.z,
        }
    }
}

impl std::ops::Add for Point {
    type Output = Point;

    fn add(self, rhs: Self) -> Self::Output {
        Point {
            x: self.x + rhs.x,
            y: self.y + rhs.y,
            z: self.z + rhs.z,
        }
    }
}

impl Point {
    pub fn zero() -> Self {
        Point { x: 0, y: 0, z: 0 }
    }

    pub fn new(coords: Vec<i32>) -> Self {
        Point {
            x: coords[0],
            y: coords[1],
            z: coords[2],
        }
    }

    pub fn squared_mag(&self) -> i32 {
        self.x * self.x + self.y * self.y + self.z * self.z
    }

    pub fn squared_dist(&self, p: &Point) -> i32 {
        let dist = self - p;
        dist.squared_mag()
    }

    pub fn manhattan(&self, p: &Point) -> i32 {
        (self.x - p.x).abs() + (self.y - p.y).abs() + (self.z - p.z).abs()
    }
}

pub struct Submarine {
    scanner_reports: Vec<Vec<Point>>,
    // Contains all the beacon positions relative to scanner 0.
    beacons_relative_to_0: BTreeSet<Point>,
    // For every `i`, contains all offsets between scanner `i` and scanner 0.
    beacon_position_relative_to_0: Vec<Point>,
}

impl Submarine {
    pub fn new() -> Self {
        Self {
            scanner_reports: Vec::new(),
            beacons_relative_to_0: BTreeSet::new(),
            beacon_position_relative_to_0: Vec::new(),
        }
    }

    pub fn on_input(&mut self, result: &str) {
        if result.starts_with("---") {
            self.scanner_reports.push(Vec::new());
            // Just initialization, we'll calculate this later.
            self.beacon_position_relative_to_0.push(Point::zero());
            return;
        }

        if result.is_empty() {
            return;
        }

        let coords: Vec<_> = result.split(",").map(|c| c.parse().unwrap()).collect();
        let point = Point::new(coords);

        self.scanner_reports.last_mut().unwrap().push(point);
    }

    fn triangulate_scanners(&mut self, i: usize, j: usize) -> bool {
        if self.beacon_position_relative_to_0[i] != Point::zero() && self.beacon_position_relative_to_0[j] != Point::zero() {
            // We already triangulated both scanners. Nothing to do, we can skip this.
            return false;
        }

        let mut common = common_beacons(&self.scanner_reports[i], &self.scanner_reports[j]);

        if common.len() < 12 {
            // Not enough beacons to triangulate. Exit.
            return false;
        }

        // `from` is the scanner that was already triangulated, `to` is the scanner we're triangulating.
        let mut from = i;
        let mut to = j;

        if self.beacon_position_relative_to_0[j] != Point::zero() {
            from = j;
            to = i;

            // Since we're switching i and j, we need to also switch the indices mapping the common beacons.
            common.iter_mut().for_each(|b| {
                *b = (b.1, b.0);
            });
        }

        // For each possible rotation, we'll check if that particular rotation puts the scanner in the coordinate system we want.
        for rotation_index in 0..ROTATIONS.len() {
            let adjusted_points: Vec<_> = self.scanner_reports[to].iter().map(|p| ROTATIONS[rotation_index](p)).collect();

            let dists: Vec<_> = common.iter().map(|&(k, l)| &self.scanner_reports[from][k] - &adjusted_points[l]).collect();
            let all_dists_equal = dists.iter().skip(1).fold(true, |s, p| s && p == &dists[0]);
            if all_dists_equal {
                // We found our rotation, so let's adjust the points from the scanner to have the correct rotation. This is important because later we'll try to triangulate this scanner with other scanners, so we need to ensure that any scanner we triangulate is in the same coordinate system as scanner 0.
                self.scanner_reports[to] = adjusted_points;

                // Figuring out the offset between this scanner and scanner 0. Important for manhattan distance later.
                self.beacon_position_relative_to_0[to] = self.beacon_position_relative_to_0[from] + dists[0];

                for p in self.scanner_reports[to].iter() {
                    self.beacons_relative_to_0.insert(*p + self.beacon_position_relative_to_0[to]);
                }

                return true;
            }
        }

        return false;
    }

    fn triangulate(&mut self) {
        let total_scanners = self.scanner_reports.len();

        // Points from scanner 0 don't need to be messed with.
        for p in &self.scanner_reports[0] {
            self.beacons_relative_to_0.insert(*p);
        }

        // Will tell us whether a given scanner was already visited and its points were already rotated.
        let mut visited = vec![false; total_scanners];
        visited[0] = true;

        // We start from scanner 0 and try to triangulate it with any other scanners.
        for i in 1..total_scanners {
            if self.triangulate_scanners(0, i) {
                visited[i] = true;
            }
        }

        loop {
            let mut visited_all = true;

            // Now we'll try to start from any visited scanner, and see if we can triangulate any unvisited scanners.
            for i in 1..total_scanners {
                if !visited[i] {
                    visited_all = false;
                    continue;
                }

                // If we're here, we already visited this scanner, so let's try to match it to any other scanners.
                for j in 1..total_scanners {
                    if visited[j] {
                        // No point in doing triangulation work between two visited scanners.
                        continue;
                    }

                    if self.triangulate_scanners(i, j) {
                        visited[j] = true;
                    }
                }
            }

            if visited_all {
                // Triangulation job is done.
                break;
            }
        }
    }

    fn max_manhattan_between_scanners(&self) -> i32 {
        let mut result = 0;

        let total_scanners = self.scanner_reports.len();

        for i in 0..total_scanners {
            for j in i + 1..total_scanners {
                let p1 = &self.beacon_position_relative_to_0[i];
                let p2 = &self.beacon_position_relative_to_0[j];

                result = result.max(p1.manhattan(p2));
            }
        }

        result
    }

    pub fn output(&mut self) {
        self.triangulate();

        println!("Part 1: {}", self.beacons_relative_to_0.len());
        println!("Part 2: {}", self.max_manhattan_between_scanners());
    }
}

// When figuring out beacons common to scanners, we need this helper struct to hold the distance between two beacons.
#[derive(Clone)]
struct PointDistanceElement {
    dist: i32,
    point_pos: usize,
}

// Following are PartialEq, Eq, PartialOrd, and Ord implementations for this helper struct. We only care about ordering based on distance. The pointer to the beacon position is important for other purposes.
impl PartialEq for PointDistanceElement {
    fn eq(&self, other: &Self) -> bool {
        self.dist == other.dist
    }
}

impl Eq for PointDistanceElement {}

impl PartialOrd for PointDistanceElement {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for PointDistanceElement {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.dist.cmp(&other.dist)
    }
}

// Given a list of all beacon positions scanned by a scanner, returns a list. Each index `i` of the returned list contains a set of all distances between beacon `i` and other beacons.
fn squared_distance_sets(scans: &Vec<Point>) -> Vec<BTreeSet<PointDistanceElement>> {
    let n = scans.len();

    let mut result = vec![BTreeSet::new(); n];

    for i in 0..n {
        for j in i + 1..n {
            let dist = scans[i].squared_dist(&scans[j]);
            // Distance between i and j is equal to distance between j and i, but the pointer position needs to be adjusted.
            result[i].insert(PointDistanceElement { dist, point_pos: j });
            result[j].insert(PointDistanceElement { dist, point_pos: i });
        }
    }

    result
}

// Given two lists of beacon positions scanned by two scanners, returns a list of mappings. Each index `i` of the returned list contains a mapping between a beacon scanned by the first scanner and a beacon scanned by the second scanner.
fn common_beacons(s1: &Vec<Point>, s2: &Vec<Point>) -> Vec<(usize, usize)> {
    let mut result = Vec::new();

    let dist1 = squared_distance_sets(s1);
    let dist2 = squared_distance_sets(s2);

    // A beacon will be common to both scanners if at least 11 other beacons from the first scanner have the same distances as 11 other beacons from the second scanner. In that case, we'll take those (at least) 12 beacons as the common beacons.
    for i in 0..dist1.len() {
        for j in 0..dist2.len() {
            let common_dists: Vec<_> = dist1[i].intersection(&dist2[j]).collect();

            // (i, j) is already a beacon, so we look for 11 or more intersections (to hit the 12 minimum beacons to triangulate).
            if common_dists.len() >= 11 {
                result.push((i, j));

                for d in common_dists {
                    // Due to how we check how a `PointDistanceElement` is equal/higher/lesser to another (we only check the distance), the intersection of distances in `common_dists` will either have a pointer to the position in the first scanner, or the position in the second scanner. To make sure that we get the proper position, we'll find the element from the intersection again in each distance set. This will ensure that the pointers will be correct.
                    let pos1 = dist1[i].get(d);
                    let pos2 = dist2[j].get(d);

                    if matches!((pos1, pos2), (Some(_), Some(_))) {
                        // Finally, we have a mapping between a beacon from the first scanner and a beacon from the second scanner.
                        result.push((pos1.unwrap().point_pos, pos2.unwrap().point_pos));
                    }
                }

                return result;
            }
        }
    }

    result
}